Fix spurious problems with unset `associatedFile`

jvican opened the pull request:
A thorough analysis and description of the issue is in the documentation
of the fix.

OlegYch committed:
add test for #559

OlegYch committed:
failing case for #559

jvican committed:
Minimize test case for #559

jvican committed:
Do best effort to detect `associatedFile` if empty

Otherwise we may end up losing internal dependencies because the
compiler failed to do a good job at setting `associatedFile` in the
classfile parser. This is what happens in #590 with default arguments,
whose test passes after this implementation.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/591/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

jvican commented:
I'd like to add some performance improvements to the implementation in case it silently regresses, so let's not merge this yet.

jvican committed:
Cache results of `findAssociatedFile` for performance

We want to make sure that if this is ever run, we don't repeat the
computation for the same class over and over again across the same
compiler run. So we install a per-cache run.

eed3si9n commented on internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala:
```diff
@@ -126,18 +126,42 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
+              // Associated file is set, so we know which classpath entry it came from
+              processExternalDependency(binaryClassName, at)
+            case None =>
+              /* If there is no associated file, it's likely the compiler didn't set it correctly.
```
Are we covering up for compiler? Should we report this upstream?

jvican replied:
In the general case, we're covering some error, but we don't know which error :smile: We've been lucky and we've found one of these errors by surfacing the bug with default arguments. I've not opened a bug report because it's not clear how to reproduce, but I've given steps to Lukas and pinged him on the bug ticket that this PR fixes so that he can clarify the situation. We'll move from there.
  
The issue here is that fixing this instance of errors is not trivial and have to be fixed bug per bug because of the way the class file parser works.

OlegYch commented:
any idea when this might be released?

eed3si9n commented:
@OlegYch it's currently in develop branch, so it will come out when sbt 1.3.0 comes out in a few months. If the patch is considered relatively stable, someone can backport it to 1.2.x branch and then it will be part of next sbt 1.2.x, which likely will come out sooner.
